### ISSUE TITLE
Lightsensors' process does not work after a few hours.

### ISSUE THREAD

**Issue by sensen963** [ryuichiueda/raspimouse_ros_2#7]
I executed a launch file in this package.
```
sengan@Raspberry-mouse:~$ roslaunch raspimouse_ros_2 raspimouse.launch  
```

After executing launch, LEDs blink and publish `/lightsensors` topics and motors rotate when I publish `/motor_raw` topics. There is no problem.
However, a few hours later, LEDs stops blinking and publishing topics.
It seems that the process of lightsensors has died (zombi process), even though the node of lightsensors seems to be alive.
I have no idea what causes this problem, ROS, my Raspberry Pi Mouse or something else.
Would you have any ideas?

before dying

processes

```
sengan@Raspberry-mouse:~$ ps -ax | grep ros
 5898 pts/19   S+     0:00 grep --color=auto ros
24448 pts/18   Sl+    0:02 /usr/bin/python /opt/ros/kinetic/bin/roslaunch raspimouse_ros_2 raspimouse.launch
24659 ?        Ssl    0:01 /usr/bin/python /opt/ros/kinetic/bin/rosmaster --core -p 11311 -w 3 __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/master.log
24773 ?        Ssl    0:00 /opt/ros/kinetic/lib/rosout/rosout __name:=rosout __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/rosout-1.log
24793 ?        Ssl    0:02 python /home/sengan/raspi_ros_tutorial/src/raspimouse_ros_2/scripts/buzzer.py __name:=buzzer __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/buzzer-2.log
24811 ?        Rsl    0:52 /home/sengan/raspi_ros_tutorial/devel/lib/raspimouse_ros_2/lightsensors __name:=lightsensors __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/lightsensors-3.log
24820 ?        Ssl    0:00 /home/sengan/raspi_ros_tutorial/devel/lib/raspimouse_ros_2/motors off __name:=motors __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/motors-4.log
24839 ?        Ssl    0:00 /home/sengan/raspi_ros_tutorial/devel/lib/raspimouse_ros_2/leds __name:=leds __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/leds-5.log
24851 ?        Rsl    0:16 /home/sengan/raspi_ros_tutorial/devel/lib/raspimouse_ros_2/buttons __name:=buttons __log:=/home/sengan/.ros/log/5bc15858-8efa-11e8-84b7-b827eb40ea72/buttons-6.log

```

nodes

```
sengan@Raspberry-mouse:~$ rosnode list
/buttons
/buzzer
/leds
/lightsensors
/motors
/rosout
```

after dying

process (zombi)
```
sengan@Raspberry-mouse:~$ ps -ax | grep lightsensors
24238 pts/19   S+     0:00 grep --color=auto lightsensors
24811 ?        Zsl    2:05 [lightsensors] <defunct>
```

node (seems to be alive)
```
sengan@Raspberry-mouse:~$ rosnode list
/buttons
/buzzer
/leds
/lightsensors
/motors
/rosout
sengan@Raspberry-mouse:~$ rosnode ping /lightsensors
rosnode: node is [/lightsensors]
pinging /lightsensors with a timeout of 3.0s
xmlrpc reply from http://Raspberry-mouse:39362/ time=2.482891ms
xmlrpc reply from http://Raspberry-mouse:39362/ time=2.837896ms
xmlrpc reply from http://Raspberry-mouse:39362/ time=1.688957ms
xmlrpc reply from http://Raspberry-mouse:39362/ time=1.724005ms
xmlrpc reply from http://Raspberry-mouse:39362/ time=1.811028ms
^Cping average: 2.108955ms

```

topic (seems to be alive but not publishes)
```
sengan@Raspberry-mouse:~$ rostopic info /lightsensors 
Type: raspimouse_ros_2/LightSensorValues

Publishers: 
 * /lightsensors (http://Raspberry-mouse:39362/)

Subscribers: None


sengan@Raspberry-mouse:~$ rostopic echo /lightsensors 
^C
```

**Comment by ryuichiueda** [comment]
> However, a few hours later, LEDs stops blinking and publishing topics.

Is a reboot of the system required after this happening? If it is needed, it would be a problem of the device driver. If you can restart the nodes with the launch file, it is a trouble of this package.

**Comment by sensen963** [comment]
Thanks for replying!

> Is a reboot of the system required after this happening?

No, no reboot is required. After killing all nodes with `Ctrl-C` in the launch's terminal, I can restart nodes with the launch file. LEDs are blinking.

In addition, when I use lightsensors without ROS (i.e. using only device driver) for a while with below codes, no problem occurs. So I do not think it is a problem of the device driver.

```
while True:
    with open('/dev/rtlightsensor0', 'r') as f:
        data = f.readline().split()
    print(data)
    time.sleep(0.1)
```

**Comment by ryuichiueda** [comment]
Thank you very much for the information. 

If I have a block of time, I would like to implement and test an exception handling for the file opening procedure. In this phase, I can't imagine other reasons. 

I also use this package for my research but I have never met this phenomenon. I would appreciate it if you could give me further information.

**Comment by sensen963** [comment]
ok.
I uses this launch file with the nodes I made. It subscribes `/lightsensors` and publishes `/motor_raw` like (https://github.com/ryuichiueda/raspimouse_lefthand/blob/master/scripts/lefthand.py).
However, the problem occurs every time in a few hours. First I thought this was because I used [
ryuichiueda/raspimouse_ros
](https://github.com/ryuichiueda/raspimouse_ros). My ROS version was kinetic and I changed the package to [ryuichiueda/raspimouse_ros_2
](https://github.com/ryuichiueda/raspimouse_ros_2). However the same problem occurs again.

While the first some hours, all nodes work correctly. 1~4 hours later some process does not work. Mostly lightsensorsdoes, but sometimes motors. (Even if /motor_raw topics flow, motors ignore them and does not change their hz.) Each time nodes remain (`rosnode echo`), but processes does not work(`ps`).

My environment is

Ubuntu version
```
sengan@Raspberry-mouse:~$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
sengan@Raspberry-mouse:~$ cat /proc/version
Linux version 4.4.38-v7+ (dc4@dc4-XPS13-9333) (gcc version 4.9.3 (crosstool-NG crosstool-ng-1.22.0-88-g8460611) ) #938 SMP Thu Dec 15 15:22:21 GMT 2016

```
ROS version
```     
 * /rosdistro: kinetic              
 * /rosversion: 1.12.7              
```

after process dying, 
printed in launch's terminal, when killing all nodes.
```
^C[buttons-6] killing on exit                 
[leds-5] killing on exit                      
[motors-4] killing on exit                    
[buzzer-2] killing on exit                    
[lightsensors-3] killing on exit              
[buttons-6] escalating to SIGTERM             
[lightsensors-3] escalating to SIGTERM        
[rosout-1] killing on exit                    
[master] killing on exit                      
shutting down processing monitor...           
... shutting down processing monitor complete 
done                                          
```

**Comment by ryuichiueda** [comment]
The engineers of RT corporation said to me that the light sensors are weak toward heat accumulation. So the sensors will have a problem if they are continuously used for some hours. Since it is only thirty minutes until the battery runs down, they may not assume that the robot is continuously used for some hours. 

Another possibility is that the device driver leaks the memory, or a variable for timer overflows. 

Anyway I'd like to observe the light sensors carefully when I use the robot.

Thank you.

**Comment by sensen963** [comment]
I am sorry for the late reply.

As you said, the cause of the problem seems heat accumulation. On the other day, I tried an experiment that I ran sensors and motors for the first 10 minutes, stopped them for the next 10 minutes and repeated that. In other words, I made them rest each 10 minutes.
Then, they continued working over 30 hours!! No processes had died.

I think heat accumulation prevented from working continuously.

I owe you a lot. Thank you.
